src/lib/pure_pursuit/PurePursuit.cpp

```diff
@@ -0,0 +1,114 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2024 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "PurePursuit.hpp"
+#include <mathlib/mathlib.h>
+
+float PurePursuit::calcDesiredHeading(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
+				      const Vector2f &curr_pos_ned,
+				      const float &lookahead_distance)
+{
+	// Check minimal necessary input validity
+	if (!PX4_ISFINITE(curr_wp_ned(0))
+	    || !PX4_ISFINITE(curr_wp_ned(1)) || !PX4_ISFINITE(curr_pos_ned(0)) || !PX4_ISFINITE(curr_pos_ned(1))) {
+		return 0.f; // No fallback possible
```

Comment:
Does this mean that we cannot command a zero heading? Wouldn't make more sense to pass a NaN in case something is wrong?

Reply:
Commanding a zero heading is not a problem.
This just catches the case where either the current waypoint or the current position are NaN (in this case I can't fallback to returning the bearing to the current waypoint).
Unfortunately the current position can become invalid at the times you wouldn't expect, for example it becomes NaN for a view update cycles after starting a mission.
I prefer returning zero here, because a NaN value can break things down the line. For example I had a slew rate in the ackermann module and if that is ever updated with a NaN value it permanently breaks.

Reply:
If you return a zero, there is no way outside of the library to detect whether if the input you have provided is valid or not.

It is actually a positive feature if something breaks down the line. It means that you need to handle these cases explicitly for your runtime application, rather than concealing it in the library.

Reply:
That is a good point, implemented in 9e00187.